### ISSUE TITLE
restic on Windows cannot backup certain symlinks

### ISSUE THREAD

**Issue by eliotmoss** [restic/restic#3614]
Output of `restic version`
--------------------------
`restic 0.12.1 (v0.11.0-657-g92412bcd) compiled with go1.17.5 on windows/amd64`
(Note: This is my own build from a repo that supplies the VSS options, which are necessary for restic to use VSS on my system.)

How did you run restic exactly?
-------------------------------
`restic backup -r $RESTIC_REPOSITORY -p C:\\path-to-passwordfile --exclude-file=C:\\path-to-exclude-file -v --use-fs-snapshot -o vss.provider=MS c:/`

`RESTIC_REPOSITORY=sftp:user@host:/path-to-remote-repo`

Contents of the exclude file:
```
c:\\$$*
c:\\DumpStack*
c:\\System*
c:\\*.sys
c:\\ProgramData\\CrashPlan\\cache*
c:\\Windows\\Temp*
```
I run this with SYSTEM user privileges, though that does not affect the error.

There are 20,000 line of output, mostly the same error.  Here is the initial portion to the first error:
```
open repository
lock repository
load index files
using parent snapshot 32848908
start scan on [c:/]
start backup on [c:/]
creating VSS snapshot for [c:\]
successfully created snapshot for [c:\]
error: open \\?\GLOBALROOT\Device\HarddiskVolumeShadowCopy7\Users\moss\AppData\Local\Microsoft\WindowsApps\91750D7E.Slack_8she8kybcnzg4\Slack.exe: The file cannot be accessed by the system.
...
```

What backend/server/service did you use to store the repository?
----------------------------------------------------------------
sftp to a server I manage

Expected behavior
-----------------
Backup symlinks.

Actual behavior
---------------
Error with the message: `The file cannot be accessed by the system.`

Some of the items are directory junctions.  Others appear to be some kind(s) of links to files/paths that are not directories.  This occurs with some standard Windows things (as illustrated above), symbolic links within an Ubuntu install for WSL (Windows Subsystem for Linux), and symbolic links within a Cygwin install.  Some of the links are to directories, some to file-like targets.

Steps to reproduce the behavior
-------------------------------
Try to backup up a Windows volume that has links of these kinds.

Do you have any idea what may have caused this?
-----------------------------------------------
Possibly related to other reports around Windows symlinks, but the error message I get is different from what I have seen given for other issues.  At least some of these links are ones that some ordinary Windows programs would not be able to follow to the target.  For example, under Cygwin, the target of one of these links will be described in terms of a path relative to the root of the Cygwin installation, not the root of the Windows volume, and will use / versus \ in the path.  I suspect the WSL Ubuntu situation is similar.  I don't know about the WindowsApps links, but they may be similarly special in the sense that they can't be directly followed, but certain programs know how to interpret the contents.  Under Cygwin some of them show as links where `ls -ld` will print the link target.  Some will not display this way, such as the entries in the Ubuntu install for `/dev/null`.  Of course those can be displayed with Ubuntu's `ls -ld`.  (My guess is that they contain specially formatted information that gives the major and minor device numbers or some such.)

Do you have an idea how to solve the issue?
-------------------------------------------
Perhaps code needs to be added to handle one or more kinds of links under Windows.  This may lie in restic itself or in Go libraries.

Did restic help you today? Did it make you happy in any way?
------------------------------------------------------------
Once I got past figuring out how to run restic with VSS and with SYSTEM privileges, it has done well, except for this one error.  It's reasonably fast to scan my nearly 3,000,000 files on a 1.1Tb volume and to send only changed data.

**Comment by eliotmoss** [comment]
A little more research shows that the Sysinternals program `junction` prints, for at least some of these links in question, `UNKNOWN MICROSOFT REPARSE POINT`.

**Comment by eliotmoss** [comment]
See also https://github.com/kmwoley/restic-windows-backup/issues/19#issue-642515490.

**Comment by rsnitsch** [comment]
I stumbled upon this issue while using WSL. Please note that I keep all my code in a certain folder on my Windows host. I can access this folder from WSL by using the automatic mounts in /mnt, e.g. /mnt/c/... to access the C: drive. Usually this works absolutely fine.

A few days ago, however, I created a catkin workspace (part of ROS) within my Windows folder. When creating a catkin workspace, catkin automatically creates some files and folders, but also one symlink, namely `/mnt/c/path/to/my/catkin_ws/src/CMakeLists.txt` pointing to `/opt/ros/noetic/share/catkin/cmake/toplevel.cmake`. Now the file `c:\path\to\my\catkin_ws\src\CMakeLists.txt` is completely inaccessible from within Windows. It can only be accessed from within WSL because it is a symlink pointing to the WSL VM's own filesystem (/opt/...).

In short: Symlinks pointing to files/folders within WSL cannot be accessed from Windows. Although you will see the file (symlink) on Windows, you simply cannot open it. And restic thus will report an error.

I am not even sure this can be fixed at all... it seems logical that the internal WSL filesystem cannot be accessed from Windows.

For me the fix was simple: Just add the file `C:\path\to\my\catkin_ws\src\CMakeLists.txt` to the restic excludes. I can exclude this file because it doesn't contain anything of value to me anyway. Alternatively you could try to backup the files in question from within WSL or whatever you're using (cygwin etc).

**Comment by yacoob** [comment]
Just bumped into this with `restic 0.15.1 compiled with go1.19.5 on windows/amd64`, for a bunch of symlinks *not pointing* to WSL locations. I wonder if it's related to [this issue](https://github.com/golang/go/issues/42184). Windows itself seems to have problems with those:

```powershell
PS C:\Users\yacoob\AppData\Local\Microsoft\WindowsApps> Get-Content .\wt.exe
Get-Content : The file cannot be accessed by the system.
At line:1 char:1
+ Get-Content .\wt.exe
+ ~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : ReadError: (C:\Users\yacoob...dowsApps\wt.exe:String) [Get-Content], IOException
    + FullyQualifiedErrorId : GetContentReaderIOError,Microsoft.PowerShell.Commands.GetContentCommand
```
